File: app/src/main/java/com/zillit/scriptdistribution/ui/sides/SidesViewModel.kt
package com.zillit.scriptdistribution.ui.sides

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.zillit.scriptdistribution.data.models.*
import com.zillit.scriptdistribution.data.repository.ScriptRepository
import kotlinx.coroutines.launch

class SidesViewModel : ViewModel() {
    private val repo = ScriptRepository()

    private val _sidesList = MutableLiveData<List<Sides>>()
    val sidesList: LiveData<List<Sides>> = _sidesList

    private val _callSheets = MutableLiveData<List<CallSheet>>()
    val callSheets: LiveData<List<CallSheet>> = _callSheets

    private val _loading = MutableLiveData(false)
    val loading: LiveData<Boolean> = _loading

    private val _downloadUrl = MutableLiveData<String?>()
    val downloadUrl: LiveData<String?> = _downloadUrl

    private val _error = MutableLiveData<String?>()
    val error: LiveData<String?> = _error

    private val _currentUser = MutableLiveData<User?>(null)
    val currentUser: LiveData<User?> = _currentUser

    /**
     * `true` when the user is allowed to create new sides (admin / editor).
     * `false` for viewer accounts — hides the Generate FAB.
     * Default permissive (true) while /auth/me is loading so the UI doesn't flash.
     */
    val canPost: LiveData<Boolean> = androidx.lifecycle.MediatorLiveData<Boolean>().apply {
        value = true
        addSource(_currentUser) { user ->
            val role = user?.role?.lowercase().orEmpty()
            value = role.isEmpty() || role != "viewer"
        }
    }

    /** Loads /api/auth/me to determine role-based UI gating. Non-fatal on failure. */
    fun loadCurrentUser() {
        viewModelScope.launch {
            try {
                val res = repo.getMe()
                if (res.isSuccessful) {
                    _currentUser.value = res.body()?.user
                }
            } catch (_: Exception) { }
        }
    }

    fun loadSides() {
        viewModelScope.launch {
            _loading.value = true
            try {
                val response = repo.listSides()
                if (response.isSuccessful) {
                    _sidesList.value = response.body()?.sides ?: emptyList()
                }
            } catch (e: Exception) {
                _error.value = e.message
            } finally {
                _loading.value = false
            }
        }
    }

    fun loadCallSheets() {
        viewModelScope.launch {
            try {
                val response = repo.listCallSheets()
                if (response.isSuccessful) {
                    _callSheets.value = response.body()?.callSheets ?: emptyList()
                }
            } catch (_: Exception) { }
        }
    }

    fun generateSides(request: GenerateSidesRequest) {
        viewModelScope.launch {
            _loading.value = true
            _error.value = null
            try {
                val response = repo.generateSides(request)
                if (response.isSuccessful) {
                    loadSides() // Refresh list
                } else {
                    _error.value = "Failed to generate sides"
                }
            } catch (e: Exception) {
                _error.value = e.message
            } finally {
                _loading.value = false
            }
        }
    }

    fun downloadSides(sidesId: String) {
        viewModelScope.launch {
            try {
                val response = repo.downloadSides(sidesId)
                if (response.isSuccessful) {
                    _downloadUrl.value = response.body()?.downloadUrl
                }
            } catch (_: Exception) { }
        }
    }

    fun deleteSides(sidesId: String) {
        viewModelScope.launch {
            try {
                repo.deleteSides(sidesId)
                loadSides()
            } catch (_: Exception) { }
        }
    }
}

File: ios/ScriptDistribution/ScriptDistribution/ViewModels/SidesViewModel.swift
//
//  SidesViewModel.swift
//
//  Drives the Sides list screen (sides + call sheets tabs + download action).
//  All async work is performed off the main thread by URLSession; `@MainActor`
//  ensures @Published mutations land on main.
//

import Foundation
import UIKit

@MainActor
final class SidesViewModel: ObservableObject {

    enum Tab: Hashable { case sides, callSheets }

    @Published var sides: [Sides] = []
    @Published var callSheets: [CallSheet] = []
    @Published var tab: Tab = .sides
    @Published var loading = false
    @Published var errorMessage: String?
    @Published var lastDownloadURL: URL?
    @Published var currentUser: User?

    private var realtimeListenerId: UUID?

    init() {
        // Refresh the list whenever the backend broadcasts a sides status change.
        realtimeListenerId = RealtimeManager.shared.addListener { [weak self] event, data in
            guard event == "sides:updated" else { return }
            Task { @MainActor [weak self] in
                guard let self = self else { return }
                await self.reload()
                if let status = data["status"] as? String, status == "error",
                   let msg = data["error"] as? String {
                    self.errorMessage = "Sides failed: \(msg)"
                }
            }
        }
    }

    deinit {
        if let id = realtimeListenerId { RealtimeManager.shared.removeListener(id) }
    }

    /// Convenience for the empty-state text.
    var isCurrentTabEmpty: Bool {
        switch tab {
        case .sides:       return sides.isEmpty
        case .callSheets:  return callSheets.isEmpty
        }
    }

    /// `true` when the user can create new sides (admin / editor / unknown role).
    /// `false` for read-only viewer accounts — hides the Generate FAB.
    /// Default is permissive (true) so the UI doesn't disappear before /auth/me resolves.
    var canPost: Bool {
        guard let role = currentUser?.role?.lowercased(), !role.isEmpty else {
            return true
        }
        return role != "viewer"
    }

    func loadInitial() async {
        // Load current user in parallel with sides so the FAB visibility settles fast.
        async let userTask: () = loadCurrentUser()
        async let dataTask: () = reload()
        _ = await (userTask, dataTask)
    }

    /// Loads /api/auth/me — used to decide whether the Generate FAB is shown.
    func loadCurrentUser() async {
        do {
            let res = try await APIService.getMe()
            currentUser = res.user
        } catch {
            // Non-fatal — leave currentUser nil and fall through to permissive default.
        }
    }

    /// Reloads whichever tab is currently selected.
    func reload() async {
        switch tab {
        case .sides:      await loadSides()
        case .callSheets: await loadCallSheets()
        }
    }

    /// Loads /api/sides.
    func loadSides() async {
        loading = true
        errorMessage = nil
        do {
            let response = try await APIService.listSides(limit: 50)
            sides = response.sides
        } catch {
            errorMessage = error.localizedDescription
        }
        loading = false
    }

    /// Loads /api/callsheets.
    func loadCallSheets() async {
        loading = true
        errorMessage = nil
        do {
            let response = try await APIService.listCallSheets(limit: 50)
            callSheets = response.callSheets
        } catch {
            errorMessage = error.localizedDescription
        }
        loading = false
    }

    /// Fetches the signed download URL for a sides record and opens it externally.
    /// Mirrors Android's `Intent.ACTION_VIEW` behaviour.
    func download(id: String) async {
        do {
            let response = try await APIService.downloadSides(id: id)
            let absoluteURL: URL?
            if response.downloadUrl.hasPrefix("/") {
                absoluteURL = URL(string: "\(APIClient.baseURL)\(response.downloadUrl)")
            } else {
                absoluteURL = URL(string: response.downloadUrl)
            }
            if let url = absoluteURL {
                await UIApplication.shared.open(url)
                lastDownloadURL = url
            } else {
                errorMessage = "Invalid download URL"
            }
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    /// Switch tabs and lazy-load the other list on first visit.
    func switchTab(to newTab: Tab) async {
        guard tab != newTab else { return }
        tab = newTab
        let isEmpty = (newTab == .sides && sides.isEmpty) || (newTab == .callSheets && callSheets.isEmpty)
        if isEmpty { await reload() }
    }
}
